pysmt/solvers/msat.py

```diff
@@ -965,16 +968,27 @@ def walk_implies(self, formula, args, **kwargs):
 
     def walk_times(self, formula, args, **kwargs):
         res = args[0]
-        nl_count = 0 if mathsat.msat_term_is_number(self.msat_env(), res) else 1
         for x in args[1:]:
-            if not mathsat.msat_term_is_number(self.msat_env(), x):
-                nl_count += 1
-            if nl_count >= 2:
-                raise NonLinearError(formula)
-            else:
-                res = mathsat.msat_make_times(self.msat_env(), res, x)
+            res = mathsat.msat_make_times(self.msat_env(), res, x)
         return res
 
+    def walk_pow(self, formula, args, **kwargs):
+        n = to_python_integer(formula.args()[1].constant_value())
+        if n == 0:
+            return mathsat.msat_make_number(self.msat_env(), "1")
+        is_neg = (n < 0)
+        n = abs(n)
+        res = args[0]
+        for i in range(2, n):
+            res = mathsat.msat_make_times(self.msat_env(), res, args[0])
+        if is_neg:
+            one = mathsat.msat_make_number(self.msat_env(), "1")
+            res = mathsat.msat_make_divide(self.msat_env(), one, res)
+        return res
+
+    def walk_div(self, formula, args, **kwargs):
+        return mathsat.msat_make_divide(self.msat_env(), args[0], args[1])
```

Comment:
Sorry if I intrude.    
If I understood correctly in pysmt `div` could be between integer or real division, while in msat `make_divide` is always real division.

We could rewrite a pysmt integer division in mathsat using floor and real division.
The following code should perform the rewriting.
```
    def walk_div(self, formula, args, **kwargs):
        if self.env.stc.get_type(formula).is_real_type():
            return mathsat.msat_make_divide(self.msat_env(), args[0], args[1])
        assert self.env.stc.get_type(formula).is_int_type()
        # smtlib2 semantics: den >= 0 ? floor(num / den) : ceil(num / den)
        # In the following we rewrite ceil(a) as -floor(-a)
        num = args[0]
        den = args[1]
        zero = mathsat.msat_make_number(self.msat_env(), "0")
        neg = mathsat.msat_make_number(self.msat_env(), "-1")
        # den >= 0
        cond = mathsat.msat_make_leq(self.msat_env(), zero, den)
        # floor(num / den)
        div = mathsat.msat_make_divide(self.msat_env(), num, den)
        div = mathsat.msat_make_floor(self.msat_env(), div)
        # - floor(num / -den)
        n_div = mathsat.msat_make_times(self.msat_env(), neg, den)
        n_div = mathsat.msat_make_divide(self.msat_env(), num, n_div)
        n_div = mathsat.msat_make_floor(self.msat_env(), n_div)
        n_div = mathsat.msat_make_times(self.msat_env(), neg, n_div)
        # den >= 0 ? floor(num / den) : - floor(num / -den)
        return mathsat.msat_make_term_ite(self.msat_env(), cond, div, n_div)
```